File: frontend/src/simulation/rules.ts
/** 任務地點規則（與 backend/app/sim/rules.py 相同）：存在、非充電樁、不相同、符合 TaskType。 */
import type { LayoutLocation } from "../layout/types";

export const TASK_RULES: Record<string, [Set<string>, Set<string>]> = {
  PICK: [new Set(["SHELF"]), new Set(["PACKING", "SORTING", "OUTBOUND"])],
  REPLENISH: [new Set(["INBOUND"]), new Set(["SHELF"])],
  TRANSPORT: [new Set(["PACKING", "SORTING", "INBOUND", "OUTBOUND"]), new Set(["PACKING", "SORTING", "INBOUND", "OUTBOUND"])],
  RETURN: [new Set(["PACKING", "SORTING", "OUTBOUND"]), new Set(["SHELF"])],
};

/** 回傳錯誤訊息；null = 合法 */
export function taskError(loc: Record<string, LayoutLocation>, type: string, source: string, destination: string): string | null {
  const rule = TASK_RULES[type];
  if (!rule) return `unknown task type ${type}`;
  const s = loc[source], d = loc[destination];
  if (!s) return `unknown source location ${source}`;
  if (!d) return `unknown destination location ${destination}`;
  if (source === destination) return "source and destination must differ";
  if (s.kind === "CHARGING" || d.kind === "CHARGING") return "charging stations cannot be task locations";
  if (!rule[0].has(s.kind)) return `${type} source must be one of ${[...rule[0]].sort().join(", ")}, got ${s.kind}`;
  if (!rule[1].has(d.kind)) return `${type} destination must be one of ${[...rule[1]].sort().join(", ")}, got ${d.kind}`;
  return null;
}

File: backend/app/sim/rules.py
"""任務地點規則（與 frontend/src/simulation/rules.ts 相同）：
source / destination 必須存在於 layout、不可是充電樁、不可相同，且依 TaskType 限制地點種類。"""
from __future__ import annotations

from typing import Any, Optional

TASK_RULES: dict[str, tuple[set[str], set[str]]] = {
    "PICK": ({"SHELF"}, {"PACKING", "SORTING", "OUTBOUND"}),
    "REPLENISH": ({"INBOUND"}, {"SHELF"}),
    "TRANSPORT": ({"PACKING", "SORTING", "INBOUND", "OUTBOUND"}, {"PACKING", "SORTING", "INBOUND", "OUTBOUND"}),
    "RETURN": ({"PACKING", "SORTING", "OUTBOUND"}, {"SHELF"}),
}


def task_error(loc: dict[str, dict[str, Any]], type_: str, source: str, destination: str) -> Optional[str]:
    """回傳錯誤訊息；None = 合法。"""
    rule = TASK_RULES.get(type_)
    if not rule:
        return f"unknown task type {type_!r}"
    s, d = loc.get(source), loc.get(destination)
    if not s:
        return f"unknown source location {source!r}"
    if not d:
        return f"unknown destination location {destination!r}"
    if source == destination:
        return "source and destination must differ"
    if s["kind"] == "CHARGING" or d["kind"] == "CHARGING":
        return "charging stations cannot be task locations"
    if s["kind"] not in rule[0]:
        return f"{type_} source must be one of {sorted(rule[0])}, got {s['kind']}"
    if d["kind"] not in rule[1]:
        return f"{type_} destination must be one of {sorted(rule[1])}, got {d['kind']}"
    return None
